QccPack-0.61-1/src/libQccPackWAV/utils/man1/sfqencode.1
.TH SFQENCODE 1 "QCCPACK" ""
.SH NAME
sfqencode, sfqdecode \- encode/decode an image using the SFQ algorithm
.SH SYNOPSIS
.B sfqencode
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-l
.IR  lambda "\|]"
.RB "[\|" \-bbq
.IR  baseband\_quantizer\_stepsize "\|]"
.RB "[\|" \-hpq
.IR  highpass\_quantizer\_stepsize "\|]"
.RB "[\|" \-pw
.IR  pcpfile "\|]"
.RB "[\|" \-nl
.IR  levels "\|]"
.RB "[\|" \-vo "\|]"
.I imgfile
.I bitstream
.LP
.B sfqdecode
.RB "[\|" \-w
.IR  wavelet "\|]"
.RB "[\|" \-b
.IR  boundary "\|]"
.RB "[\|" \-pw
.IR  pcpfile "\|]"
.I bitstream
.I imgfile
.SH OPTIONS
.TP
.BI \-w " wavelet"
String. 
Name of the wavelet file
.RB ( FBK
or
.B LFT
format); default = "CohenDaubechiesFeauveau.9-7.lft"
.TP
.BI \-b " boundary"
String. Type of boundary extension; default = "symmetric".
.TP
.BI \-l " lambda"
Float. Rate-distortion parameter.  Default is 1.0.
.TP
.BI \-bbq " baseband\_quantizer\_stepsize"
Float.  Default is to search all allowable stepsizes.
.TP
.BI \-hpq " highpass\_quantizer\_stepsize"
Float.  Default is to search all allowable stepsizes.
.TP
.BI \-pw " pcpfile
String. Name of file containing perceptual weights.
.TP 
.BI \-nl " levels"
Integer. Number of levels of dyadic decomposition.
Default value is 3 levels.
.TP
.B \-vo
Print only the value of the rate to stdout (terse output).
Default is verbose output.
.SH DESCRIPTION
.LP
.B sfqencode
codes an image,
.IR imgfile ,
using the space-frequency-quantization (SFQ) algorithm of Xiong
.IR "et al" .
The SFQ algorithm involves a 2D DWT followed by the
identification of zerotrees based on a rate-distortion criterion.
The zerotree structure indicates which wavelet coefficients are 
"insignificant."
Coding these
insignificant coefficients costs more in bit rate
than the possible gain in distortion performance;
consequently, these insignificant coefficients are set to zero.
Uniform scalar quantization 
followed by arithmetic entropy coding
is used to code all remaining significant coefficients.
.LP
The DWT is performed as described for
.BR imgdwt (1).
Option
.B \-w
specifies the wavelet, 
option
.B \-b
specifies the boundary extension, and option
.B \-pw
indicates whether perceptual weighting is to be used.
See 
.BR imgdwt (1)
for more details.
.LP
The SFQ algorithm operates using a Lagrangian cost formulation
of the form
.I "J = D + lambda * R"
where 
.I D
is distortion,
.I R
is rate,
and
.I lambda
is the Lagrangian rate-distortion parameter given by
the
.B \-l
option.
.I lambda 
determines the balance between rate and distortion in the
algorithm:  large values of
.I lambda
emphasize reduction of bit rate over distortion; small values of
.I lambda
produce codings with lower distortion but at a consequently higher rate.
Varying
.I lambda
from 0 to infinity will trace out the rate-distortion performance curve
of the SFQ algorithm for a particular input image and for a
particular scalar-quantizer stepsize; a single
.I lambda
thus specifies an operating point on this rate-distortion curve.
.LP
The performance of the SFQ algorithm depends not only on 
.I lambda
but also on the chosen scalar-quantizer stepsizes.
The option
.B \-hpq
specifies the scalar-quantizer stepsize to be used in the coding
of the highpass subbands; the baseband, coded independently from the
highpass subbands, uses the stepsize given by the
.B \-bbq
option.
If a highpass stepsize is not specified (i.e., no
.B \-hpq
option given), 
.B sfqencode
determines an "optimal" stepsize by searching in
a set of 245 stepsizes between 7.6 and 32.
The optimal stepsize is the one in the set found to
minimize the rate-distortion cost criterion
.I J
(see above).
Note: this search process essentially implies coding the highpass
subbands 245 times and
choosing the best coding; it is consequently quite slow.
A similar search is performed for the baseband stepsize in the
case that the
.B \-bbq
option is not given.
.LP
.B sfqencode
normally prints to stdout a small amount of information
concerning the encoding being performed, most notably the bit rate,
expressed in bits per pixel,
achieved by the program.  If the 
.B \-vo
(value-only) option is given, this printed output is suppressed except for
the bit rate.
.LP
Note: the original presentation of SFQ given by Xiong 
.IR "et al" .
describes a method for predicting portions of the zerotree for
high-frequency subbands based on the zerotree for
lower-frequency subbands.
.BR sfqencode ,
as currently implemented, does not perform this zerotree prediction.
.LP
Note: the original presentation of SFQ given by Xiong
.IR "et al" .
does not include perceptual weighting as allowed by
.B sfqencode
(the 
.B \-pw
option).  Refer to
.BR imgdwt (1)
and
.BR QccWAVPerceptualWeights (3)
for details on the implementation of perceptual weighting.
.LP
Note:
.BR sfqencode
currently achieves rate-distortion performance slightly below that reported
by Xiong
.IR "et al" .
For instance, when coding the "Barbara" 
image at 0.5 bits/pixel, 
.B sfqencode 
achieves a PSNR about 1dB below the reported figure.
The reasons for this performance gap are currently being investigated,
although it is probably due in part to the lack of tree prediction
as described above.
.SH "SEE ALSO"
.BR imgdwt (1),
.BR QccWAVPerceptualWeights (3),
.BR QccWAVSubbandPyramid (3),
.BR QccPackWAV (3),
.BR QccPackIMG (3),
.BR QccPack (3)

Z. Xiong, K. Ramchandran, and M. T. Orchard,
"Space-Frequency Quantization for Wavelet Image Coding,"
.IR "IEEE Transactions on Image Processing" ,
vol. 6, no. 5, pp. 677-693, May 1997.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackWAV/utils/sfqencode.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include "sfqencode.h"

#define USG_STRING "[-w %s:wavelet] [-b %s:boundary] [-l %f:lambda] [-bbq %: %f:baseband_quantizer_stepsize] [-hpq %: %f:highpass_quantizer_stepsize] [-pw %: %s:pcpfile] [-nl %d:num_levels] [-vo %:] %s:imgfile %s:bitstream"

QccWAVWavelet Wavelet;
QccString WaveletFilename = QCCWAVWAVELET_DEFAULT_WAVELET;
QccString Boundary = "symmetric";

QccWAVPerceptualWeights PerceptualWeights;
int UsePerceptualWeights = 0;

float Lambda = 1.0;
int NumLevels = 3;
int NumSubbands;

int BasebandQuantizerStepSizeSpecified = 0;
int HighpassQuantizerStepSizeSpecified = 0;
float InputBasebandQuantizerStepSize = -1.0;
float InputHighpassQuantizerStepSize = -1.0;
double BasebandQuantizerStepSize;
double HighpassQuantizerStepSize;

QccIMGImage InputImage;
int ImageNumRows, ImageNumCols;

QccSQScalarQuantizer MeanQuantizer;
QccSQScalarQuantizer BasebandQuantizer;
QccSQScalarQuantizer HighpassQuantizer;

QccBitBuffer OutputBuffer;

QccWAVZerotree Zerotree;
QccChannel *Channels;

int ValueOnly = 0;

/*  Tree prediction is not working -- do not use!!  */
int NoTreePrediction = 1;
int *HighTreePredictionThresholds = NULL;
int *LowTreePredictionThresholds = NULL;

QccWAVSubbandPyramid SubbandPyramid;

QccIMGImageComponent ReconstructedBaseband;

int main(int argc, char *argv[])
{
  int subband;
  double rate_header, rate_zerotree, rate_channels;
  double rate;

  QccInit(argc, argv);
  
  QccWAVWaveletInitialize(&Wavelet);
  QccWAVPerceptualWeightsInitialize(&PerceptualWeights);
  QccIMGImageInitialize(&InputImage);
  QccSQScalarQuantizerInitialize(&MeanQuantizer);
  QccSQScalarQuantizerInitialize(&BasebandQuantizer);
  QccSQScalarQuantizerInitialize(&HighpassQuantizer);
  QccWAVZerotreeInitialize(&Zerotree);
  QccWAVSubbandPyramidInitialize(&SubbandPyramid);
  QccIMGImageComponentInitialize(&ReconstructedBaseband);
  QccBitBufferInitialize(&OutputBuffer);

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         WaveletFilename,
                         Boundary,
                         &Lambda,
                         &BasebandQuantizerStepSizeSpecified,
                         &InputBasebandQuantizerStepSize,
                         &HighpassQuantizerStepSizeSpecified,
                         &InputHighpassQuantizerStepSize,
                         &UsePerceptualWeights,
                         PerceptualWeights.filename,
                         &NumLevels,
                         &ValueOnly,
                         InputImage.filename,
                         OutputBuffer.filename))
    QccErrorExit();
  
  if (Lambda < 0)
    {
      QccErrorAddMessage("%s: Lambda must be nonnegative",
                         argv[0]);
      QccErrorExit();
    }
  
  if (NumLevels < 0)
    {
      QccErrorAddMessage("%s: Number of levels of decomposition must be nonnegative",
                         argv[0]);
      QccErrorExit();
    }
  NumSubbands = QccWAVSubbandPyramidNumLevelsToNumSubbands(NumLevels);
  
  if (BasebandQuantizerStepSizeSpecified)
    if (InputBasebandQuantizerStepSize <= 0)
      {
        QccErrorAddMessage("%s: Baseband-quantizer stepsize must be greater than 0",
                           argv[0]);
        QccErrorExit();
      }
  BasebandQuantizerStepSize = InputBasebandQuantizerStepSize;

  if (HighpassQuantizerStepSizeSpecified)
    if (InputHighpassQuantizerStepSize <= 0)
      {
        QccErrorAddMessage("%s: Highpass-quantizer stepsize must be greater than 0",
                           argv[0]);
        QccErrorExit();
      }
  HighpassQuantizerStepSize = InputHighpassQuantizerStepSize;

  if (QccWAVWaveletCreate(&Wavelet, WaveletFilename, Boundary))
    {
      QccErrorAddMessage("%s: Error calling QccWAVWaveletCreate()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (UsePerceptualWeights)
    {
      if (QccWAVPerceptualWeightsRead(&PerceptualWeights))
        {
          QccErrorAddMessage("%s: Error calling QccWAVPerceptualWeightsRead()",
                             argv[0]);
          QccErrorExit();
        }
      if (PerceptualWeights.num_subbands != NumSubbands)
        {
          QccErrorAddMessage("%s: Number of subbands in %s does not match specified number of levels (%d)",
                             argv[0],
                             PerceptualWeights.filename,
                             NumLevels);
        }
    }

  MeanQuantizer.type = QCCSQSCALARQUANTIZER_UNIFORM;
  MeanQuantizer.stepsize = QCCWAVSFQ_MEANQUANTIZER_STEPSIZE;
  
  if (QccIMGImageRead(&InputImage))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageRead()",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageColor(&InputImage))
    {
      QccErrorAddMessage("%s: Color images not supported",
                         argv[0]);
      QccErrorExit();
    }
  if (QccIMGImageGetSize(&InputImage,
                         &ImageNumRows, &ImageNumCols))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageGetSize()",
                         argv[0]);
      QccErrorExit();
    }
  if (ImageNumCols != ((int)(ImageNumCols >> NumLevels) << NumLevels))
    {
      QccErrorAddMessage("%s: Image size (%d x %d) is not a multiple of %d (as needed for a decomposition of %d levels)",
                         argv[0],
                         ImageNumCols, ImageNumRows,
                         (1 << NumLevels), NumLevels);
      QccErrorExit();
    }
  if (ImageNumRows != ((int)(ImageNumRows >> NumLevels) << NumLevels))
    {
      QccErrorAddMessage("%s: Image size (%d x %d) is not a multiple of %d (as needed for a decomposition of %d levels",
                         argv[0],
                         ImageNumCols, ImageNumRows,
                         (1 << NumLevels), NumLevels);
      QccErrorExit();
    }
  
  if ((Channels = 
       (QccChannel *)malloc(sizeof(QccChannel)*NumSubbands)) == NULL)
    {
      QccErrorAddMessage("%s: Error allocating memory",
                         argv[0]);
      QccErrorExit();
    }
  
  Zerotree.image_num_cols = ImageNumCols;
  Zerotree.image_num_rows = ImageNumRows;
  Zerotree.num_levels = NumLevels;
  Zerotree.num_subbands = NumSubbands;
  
  if (QccWAVZerotreeMakeFullTree(&Zerotree))
    {
      QccErrorAddMessage("%s: Error calling QccWAVZerotreeMakeFullTree()",
                         argv[0]);
      QccErrorExit();
    }
  
  for (subband = 0; subband < NumSubbands; subband++)
    {
      QccChannelInitialize(&Channels[subband]);
      Channels[subband].channel_length = 
        Zerotree.num_rows[subband] * Zerotree.num_cols[subband];
      Channels[subband].access_block_size = 
        QCCCHANNEL_ACCESSWHOLEFILE;
      if (QccChannelAlloc(&Channels[subband]))
        {
          QccErrorAddMessage("%s: Error calling QccChannelAlloc()",
                             argv[0]);
          QccErrorExit();
        }
    }
      
  BasebandQuantizer.type = QCCSQSCALARQUANTIZER_UNIFORM;
  BasebandQuantizer.stepsize = BasebandQuantizerStepSize;

  HighpassQuantizer.type = QCCSQSCALARQUANTIZER_UNIFORM;
  HighpassQuantizer.stepsize = HighpassQuantizerStepSize;


  if (!NoTreePrediction)
    {
      if ((HighTreePredictionThresholds =
           (int *)malloc(sizeof(int) *
                         NumSubbands)) == NULL)
        {
          QccErrorAddMessage("%s: Error allocating memory",
                             argv[0]);
          QccErrorExit();
        }
      if ((LowTreePredictionThresholds =
           (int *)malloc(sizeof(int) *
                         NumSubbands)) == NULL)
        {
          QccErrorAddMessage("%s: Error allocating memory",
                             argv[0]);
          QccErrorExit();
        }
    }

  SubbandPyramid.num_levels = 0;
  SubbandPyramid.num_rows = InputImage.Y.num_rows;
  SubbandPyramid.num_cols = InputImage.Y.num_cols;
  if (QccWAVSubbandPyramidAlloc(&SubbandPyramid))
    {
      QccErrorAddMessage("%s: Error calling QccWAVSubbandPyramidAlloc()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccWAVsfqWaveletAnalysis(&(InputImage.Y),
                               &SubbandPyramid,
                               &Zerotree,
                               &MeanQuantizer,
                               &Wavelet,
                               ((UsePerceptualWeights) ? &PerceptualWeights :
                                NULL),
                               (double)Lambda))
    {
      QccErrorAddMessage("%s: Error calling QccWAVsfqWaveletAnalysis()",
                         argv[0]);
      QccErrorExit();
    }

  ReconstructedBaseband.num_rows = Zerotree.num_rows[0];
  ReconstructedBaseband.num_cols = Zerotree.num_cols[0];
  if (QccIMGImageComponentAlloc(&ReconstructedBaseband))
    {
      QccErrorAddMessage("%s: Error calling QccIMGImageComponentAlloc()",
                         argv[0]);
      QccErrorExit();
    }

  if (QccWAVsfqBasebandEncode(&SubbandPyramid,
                              &Zerotree,
                              &BasebandQuantizer,
                              Channels,
                              (double)Lambda,
                              &ReconstructedBaseband))
    {
      QccErrorAddMessage("%s: Error calling QccWAVsfqBasebandEncode()",
                         argv[0]);
      QccErrorExit();
    }
  if (!ValueOnly)
    printf("Baseband quantizer: chosen stepsize = %f\n",
           BasebandQuantizer.stepsize);

  if (QccWAVsfqHighpassEncode(&SubbandPyramid,
                              &Zerotree,
                              &HighpassQuantizer,
                              Channels,
                              (double)Lambda,
                              HighTreePredictionThresholds,
                              LowTreePredictionThresholds,
                              &ReconstructedBaseband))
    {
      QccErrorAddMessage("%s: Error calling QccWAVsfqHighpassEncode()",
                         argv[0]);
      QccErrorExit();
    }

  if (!ValueOnly)
    printf("Highpass quantizer: chosen stepsize = %f\n",
           HighpassQuantizer.stepsize);

  OutputBuffer.type = QCCBITBUFFER_OUTPUT;
  if (QccBitBufferStart(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferStart()",
                         argv[0]);
      QccErrorExit();
    }
  
  if (QccWAVsfqAssembleBitstream(&Zerotree,
                                 &MeanQuantizer,
                                 &BasebandQuantizer,
                                 &HighpassQuantizer,
                                 Channels,
                                 HighTreePredictionThresholds,
                                 LowTreePredictionThresholds,
                                 &OutputBuffer,
                                 &rate_header,
                                 &rate_zerotree,
                                 &rate_channels))
    {
      QccErrorAddMessage("%s: Error calling QccWAVsfqAssembleBitstream()",
                         argv[0]);
      QccErrorExit();
    }
      
  if (QccBitBufferEnd(&OutputBuffer))
    {
      QccErrorAddMessage("%s: Error calling QccBitBufferEnd()",
                         argv[0]);
      QccErrorExit();
    }

  rate = rate_header + rate_zerotree + rate_channels;

  if (ValueOnly)
    printf("%f\n", rate);
  else
    {
      printf("SFQ coding of %s:\n",
             InputImage.filename);
      printf("  Total rate: %f bpp\n", rate);
      printf("      Header: %6.4f bpp (% 6.2f%% of total rate)\n",
             rate_header, QccMathPercent(rate_header, rate));
      printf("    Zerotree: %6.4f bpp (% 6.2f%% of total rate)\n",
             rate_zerotree, QccMathPercent(rate_zerotree, rate));
      printf("    Channels: %6.4f bpp (% 6.2f%% of total rate)\n",
             rate_channels, QccMathPercent(rate_channels, rate));
    }
  
  QccWAVWaveletFree(&Wavelet);
  QccWAVPerceptualWeightsFree(&PerceptualWeights);
  QccIMGImageFree(&InputImage);
  QccSQScalarQuantizerFree(&MeanQuantizer);
  QccSQScalarQuantizerFree(&BasebandQuantizer);
  QccSQScalarQuantizerFree(&HighpassQuantizer);
  QccWAVZerotreeFree(&Zerotree);
  QccWAVSubbandPyramidFree(&SubbandPyramid);
  QccIMGImageComponentFree(&ReconstructedBaseband);

  QccExit;
}
